Abt report

NoahCarnahan opened the pull request:
Adds a custom UCR base item expression for a custom report.

This also adds the report configurations to the repository, but they don't necessarily need to be there. I wanted to keep them under source control, and this seemed like the easiest way, but I can remove them before merging if desirable.

NoahCarnahan committed:
temp

NoahCarnahan committed:
moar temp

NoahCarnahan committed:
save report source here for now in case of data loss

NoahCarnahan committed:
update custom base item expression

NoahCarnahan committed:
Update the way in which flags are specified

NoahCarnahan committed:
move report configuration backups

NoahCarnahan committed:
move abt_supervisor expression out of core

NoahCarnahan committed:
report config updates

NoahCarnahan committed:
specify flags in yaml

NoahCarnahan committed:
Add more flag specs

NoahCarnahan committed:
update report configs

NoahCarnahan committed:
added inspector names to supervisor report

NoahCarnahan committed:
Added sms report

NoahCarnahan committed:
update flag specs

NoahCarnahan committed:
update flagspecs with some easy logic questions

NoahCarnahan committed:
Add logic for unchecked warning type

NoahCarnahan committed:
Added support for repeats and from2 flag specs

NoahCarnahan committed:
update report configs

NoahCarnahan committed:
Pull flag comments from submitted form

NoahCarnahan committed:
updated things

NoahCarnahan committed:
translations

NoahCarnahan committed:
more french

NoahCarnahan committed:
Various fixes

NoahCarnahan committed:
French for form 2

NoahCarnahan committed:
French for form 1

NoahCarnahan committed:
fix indenets

NoahCarnahan committed:
Show french warnings if country is francophone

NoahCarnahan committed:
Add missing flags to form 3 spec

NoahCarnahan committed:
Add form 3 translations

NoahCarnahan committed:
fix double q5_2

NoahCarnahan committed:
Add logic for special case question

NoahCarnahan committed:
Get inspector names from form submissions

NoahCarnahan committed:
add spray progress and coverage to sms report

NoahCarnahan committed:
Update report column headings to french

NoahCarnahan committed:
update column order

dimagimon commented on custom/abt/reports/expressions.py:
```diff
@@ -1,15 +1,215 @@
+
+
+
+        names = []
```
E303 too many blank lines (3)

NoahCarnahan committed:
remove extra new lines

czue commented on custom/abt/reports/progress_report.json:
```diff
@@ -0,0 +1 @@
+{"doc_type": "ReportConfiguration", "domain": "abt-test-2", "config_id": "0ff37d5a88231f873d4d98b410a2c3c0", "description": "", "title": "Spray Progress Report", "filters": [{"field": "date_of_data_collection", "required": false, "type": "date", "slug": "date_of_data_collection", "display": "Date of data collection"}, {"required": false, "datatype": "string", "show_all": true, "display": "Pays",  ...
```
might be worth prettifying this for future edits and readability

czue commented on custom/abt/reports/sms_report.json:
```diff
@@ -0,0 +1 @@
+{"doc_type": "ReportConfiguration", "domain": "abt-test-2", "config_id": "0ff37d5a88231f873d4d98b410a2c3c0", "description": "", "title": "SMS Indicator Report", "filters": [{"required": false, "datatype": "string", "show_all": true, "display": "Pays", "field": "country", "type": "dynamic_choice_list", "slug": "country"}, {"required": false, "datatype": "string", "show_all": true, "display": "Nive ...
```
(all of them)

czue commented on custom/abt/reports/expressions.py:
```diff
@@ -1,15 +1,213 @@
+        (which is a dict representation of an XFormInstance)
+        """
+        app_id, xmlns = item['app_id'], item['xmlns']
+        questions = cls._questions_cache.get((app_id, xmlns), None)
```
given that this is a classmethod I think this cache will live for essentially the lifecycle of the process, right? on the one hand that's great from a caching/speed perspective, though it may have unexpected behavior:
1. memory-wise - are there lotso of app_id/xmlns pairs? if so this may grow to be a high footprint attached to every python process.
2. if you change the app, the cache won't change until the process restarts (e.g. gunicorn serves a certain number of requests, pillowtop restarts, there's a deploy, etc.). I think that's fine since app changes might require table rebuilds anyways, just noting.

generally am interested how you made this design decision as opposed to using something much simpler like `quickcache`

NoahCarnahan replied:
Ah thanks for pointing that out. I hadn't considered how long that would stay around. There are only 7 app_id/xmlns pairs, so not a huge deal. I went with this approach because I forgot that `quickcache` exists :) I'll convert it to use `quickcache` now.

NoahCarnahan committed:
prettify json

czue commented:
I searched your table definitions for 'abt_supervisor' and didn't find anything. I think this is probably a dumb question, but where is the custom expression actually being used?

czue commented:
This looks great overall - I reviewed with a similar bar as custom report code (so not too closely) but the code looks great, super clean and very nicely documented!

NoahCarnahan commented:
Ha yeah. The custom expression is used in the data source "base item
expression" field, which I didn't actually put under source control. Since
the UCR interface doesn't have a "datasource source" view like the "Report
source" view, I just copied and pasted the long parts, of the
configuration, which are the indicators, so there is still a little bit of
clicking I need to do to set up the report on prod (including entering
"abt_supervisor" in the appropriate field).

On Wed, May 20, 2015 at 2:36 PM, Cory Zue <notifications@github.com> wrote:

> I searched your table definitions for 'abt_supervisor' and didn't find
> anything. I think this is probably a dumb question, but where is the custom
> expression actually being used?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/6761#issuecomment-103852831>.
>

czue commented on custom/abt/reports/supervisory_datasource_indicators.json:
```diff
@@ -0,0 +1,225 @@
+[
```
How are these indicator files used? You may be interested in `CustomDataSourceConfiguration`s which I think might be designed for this use case? Essentially a config file you keep in the repo that will automatically sync to a UCR table.

NoahCarnahan replied:
Oh awesome, I didn't know that existed. I was copy and pasting the indicator files. I put them in the repo incase I wanted to review the changes. Is there documentation on how to use `CustomDataSourceConfiguration`s?

czue replied:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/README.md#static-data-sources

czue replied:
everything in UCR has docs :)

NoahCarnahan replied:
:heart_eyes: the UCR docs

czue commented:
Gotcha. You should just make a "data source source" view. :) Or you can grab the json straight out of couch.

NoahCarnahan committed:
use quickcache instead of one off solution

NoahCarnahan commented:
@czue 
Thanks for the feedback. I think I've addressed all your comments except using `CustomDataSourceConfiguration`s. I gave the `CustomDataSourceConfiguration` a try but I wasn't able to select data sources created this way on the report edit page. I'm guessing it's an easy fix, but I'm a bit anxious to get this all deployed and set up by Friday (when the report is needed). So, if we could merge this as is, then switch to the `CustomDataSourceConfiguration` in the coming days that would be great.

czue commented:
@NoahCarnahan sounds good. I'll merge when tests pass.

NoahCarnahan commented:
Convert to `CustomDataSourceConfiguration` here: https://github.com/dimagi/commcare-hq/pull/7013
